mypyc/irbuild/ll_builder.py

```diff
@@ -588,7 +588,17 @@ def compare_tagged(self, lhs: Value, rhs: Value, op: str, line: int) -> Value:
         branch.negated = False
         self.add(branch)
         self.activate_block(short_int_block)
-        eq = self.binary_int_op(bool_rprimitive, lhs, rhs, op_type, line)
+        # for now equal op, explicitly convert tagged(unsigned) to signed to include negative
+        # situtations
+        if op not in ("==", "!="):
+            lhs_short = self.add(Truncate(lhs, short_int_rprimitive,
+                                          c_pyssize_t_rprimitive, line))
+            rhs_short = self.add(Truncate(rhs, short_int_rprimitive,
+                                          c_pyssize_t_rprimitive, line))
```

Comment:
This is not very principled. The truncate op is redundant, and also confusingly the sizes of source and target types are the same, so there's no truncation actually going on. 

Instead, we should generate the correct C code for short int comparison (add the cast to a signed integer type). Maybe you can look at the `type` attribute of `BinaryIntOp` to determine whether a signed comparison is needed when emitting C.

Reply:
Yeah, the truncate also looks weird to me. So when we emit a BinaryIntOp, we check if both operands are short_int(and probably check if the op is not EQ or NEQ as well since they do not require casts to have the correct result) to cast them into `py_ssize_t`

Reply:
More generally, also plain tagged integers need to use a cast since they are signed. 

Probably the cleanest way to check the operand types would be to update every `Value` to contain the result type, but I'm not sure if this is necessary.